File: macos/Sources/BridgeyMac/GamingUdpFraming.swift
import Foundation

/// GAMING MODE POC (not wired into production capture/decode/input paths). Swift mirror of
/// GamingUdpFraming.kt - see that file's doc comment for the full wire-format rationale (AAD-bound
/// header, MTU-safe fragmentation, why this deliberately improves on M1's TCP framing where
/// kind/sessionId aren't AEAD-bound).
enum GamingUdpFraming {
    static let version: UInt8 = 1
    static let headerBytes = 1 + 1 + 8 + 8 + 2 + 2 + 8 + 1 + 4
    static let nonceBytes = 12
    static let tagBytes = 16
    static let maxFragmentPayloadBytes = 1200

    enum PacketType {
        static let videoFrameFragment: UInt8 = 0
        static let inputMove: UInt8 = 1
        static let heartbeat: UInt8 = 2
        static let all: Set<UInt8> = [videoFrameFragment, inputMove, heartbeat]
    }

    enum Flags {
        static let keyframe = 1
        static let config = 2
    }

    struct Header {
        let type: UInt8
        let gamingSessionToken: UInt64
        let frameId: UInt64
        let fragmentIndex: Int
        let fragmentCount: Int
        let captureTimestampMs: UInt64
        let flags: Int
        let sequence: UInt32
    }

    struct ParsedPacket {
        let header: Header
        let aad: Data
        let nonce: Data
        let ciphertext: Data
    }

    static func encodeHeader(_ header: Header) -> Data {
        var data = Data(capacity: headerBytes)
        data.append(version)
        data.append(header.type)
        data.append(contentsOf: header.gamingSessionToken.bigEndianBytes)
        data.append(contentsOf: header.frameId.bigEndianBytes)
        data.append(contentsOf: UInt16(header.fragmentIndex).bigEndianBytes)
        data.append(contentsOf: UInt16(header.fragmentCount).bigEndianBytes)
        data.append(contentsOf: header.captureTimestampMs.bigEndianBytes)
        data.append(UInt8(header.flags))
        data.append(contentsOf: header.sequence.bigEndianBytes)
        return data
    }

    static func parse(_ packet: Data) -> ParsedPacket? {
        guard packet.count >= headerBytes + nonceBytes + tagBytes else { return nil }
        let bytes = [UInt8](packet)
        guard bytes[0] == version else { return nil }
        let type = bytes[1]
        guard PacketType.all.contains(type) else { return nil }
        var offset = 2
        let gamingSessionToken = UInt64(bigEndianBytes: bytes, at: &offset)
        let frameId = UInt64(bigEndianBytes: bytes, at: &offset)
        let fragmentIndex = Int(UInt16(bigEndianBytes: bytes, at: &offset))
        let fragmentCount = Int(UInt16(bigEndianBytes: bytes, at: &offset))
        guard fragmentCount > 0, fragmentIndex < fragmentCount else { return nil }
        let captureTimestampMs = UInt64(bigEndianBytes: bytes, at: &offset)
        let flags = Int(bytes[offset]); offset += 1
        let sequence = UInt32(bigEndianBytes: bytes, at: &offset)
        let aad = packet.subdata(in: 0..<headerBytes)
        let nonce = packet.subdata(in: headerBytes..<(headerBytes + nonceBytes))
        let ciphertext = packet.subdata(in: (headerBytes + nonceBytes)..<packet.count)
        return ParsedPacket(
            header: Header(
                type: type, gamingSessionToken: gamingSessionToken, frameId: frameId,
                fragmentIndex: fragmentIndex, fragmentCount: fragmentCount,
                captureTimestampMs: captureTimestampMs, flags: flags, sequence: sequence
            ),
            aad: aad, nonce: nonce, ciphertext: ciphertext
        )
    }

    static func assemble(_ header: Header, nonce: Data, ciphertext: Data) -> Data {
        encodeHeader(header) + nonce + ciphertext
    }

    static func splitIntoFragments(_ payload: Data) -> [Data] {
        guard !payload.isEmpty else { return [Data()] }
        var fragments: [Data] = []
        var offset = 0
        while offset < payload.count {
            let end = min(offset + maxFragmentPayloadBytes, payload.count)
            fragments.append(payload.subdata(in: offset..<end))
            offset = end
        }
        return fragments
    }
}

private extension FixedWidthInteger {
    var bigEndianBytes: [UInt8] {
        withUnsafeBytes(of: bigEndian, Array.init)
    }
}

private extension UInt64 {
    init(bigEndianBytes bytes: [UInt8], at offset: inout Int) {
        var value: UInt64 = 0
        for i in 0..<8 { value = (value << 8) | UInt64(bytes[offset + i]) }
        offset += 8
        self = value
    }
}

private extension UInt32 {
    init(bigEndianBytes bytes: [UInt8], at offset: inout Int) {
        var value: UInt32 = 0
        for i in 0..<4 { value = (value << 8) | UInt32(bytes[offset + i]) }
        offset += 4
        self = value
    }
}

private extension UInt16 {
    init(bigEndianBytes bytes: [UInt8], at offset: inout Int) {
        var value: UInt16 = 0
        for i in 0..<2 { value = (value << 8) | UInt16(bytes[offset + i]) }
        offset += 2
        self = value
    }
}

File: android/app/src/main/java/dev/bridgey/android/GamingUdpFraming.kt
package dev.bridgey.android

import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * GAMING MODE POC (not wired into production capture/decode/input paths - see
 * GamingModePoc.md investigation notes). Defines the UDP realtime packet format used to prove out
 * fragmentation/reassembly/security mechanics before any production Gaming Mode is built.
 *
 * Wire format (all integers big-endian):
 *   [1B version][1B type][8B gamingSessionToken][8B frameId][2B fragmentIndex][2B fragmentCount]
 *   [8B captureTimestampMs][1B flags][4B sequence][12B nonce][ciphertext + 16B GCM tag]
 *
 * Everything before the nonce is the AEAD's "associated data" (AAD) - authenticated but not
 * encrypted - so frameId/fragmentIndex/flags/sequence/etc. cannot be tampered with by a network
 * observer without invalidating the GCM tag, even though they're readable in plaintext for routing
 * (deciding which reassembly bucket a fragment belongs to, or dropping stale packets) before
 * spending a decrypt call. This is a deliberate improvement over the frozen M1 TCP framing (there,
 * `kind`/`sessionId` travel outside any AEAD binding entirely - acceptable there only because every
 * actionable command additionally requires a successful decrypt of its own payload as proof of
 * session possession; here we can do better for free since AAD support costs nothing extra).
 */
internal object GamingUdpFraming {
    const val VERSION: Byte = 1
    const val HEADER_BYTES = 1 + 1 + 8 + 8 + 2 + 2 + 8 + 1 + 4
    const val NONCE_BYTES = 12
    const val TAG_BYTES = 16

    /** Safely below the standard 1500B Ethernet MTU after IP/UDP/our-header overhead, avoiding IP
     *  fragmentation (which would reintroduce exactly the reliability dependency UDP is meant to
     *  avoid here - a single lost IP fragment would silently corrupt the whole UDP datagram). */
    const val MAX_FRAGMENT_PAYLOAD_BYTES = 1200

    object PacketType {
        const val VIDEO_FRAME_FRAGMENT: Byte = 0
        const val INPUT_MOVE: Byte = 1
        const val HEARTBEAT: Byte = 2
        val all = setOf(VIDEO_FRAME_FRAGMENT, INPUT_MOVE, HEARTBEAT)
    }

    object Flags {
        const val KEYFRAME: Int = 1
        const val CONFIG: Int = 2
    }

    data class Header(
        val type: Byte,
        val gamingSessionToken: Long,
        val frameId: Long,
        val fragmentIndex: Int,
        val fragmentCount: Int,
        val captureTimestampMs: Long,
        val flags: Int,
        val sequence: Int,
    )

    data class ParsedPacket(val header: Header, val aad: ByteArray, val nonce: ByteArray, val ciphertext: ByteArray)

    /** Builds the plaintext header bytes (used both as the wire prefix and as AEAD associated
     *  data) - the caller encrypts `payload` separately and appends nonce+ciphertext+tag. */
    fun encodeHeader(header: Header): ByteArray {
        val buffer = ByteBuffer.allocate(HEADER_BYTES).order(ByteOrder.BIG_ENDIAN)
        buffer.put(VERSION)
        buffer.put(header.type)
        buffer.putLong(header.gamingSessionToken)
        buffer.putLong(header.frameId)
        buffer.putShort(header.fragmentIndex.toShort())
        buffer.putShort(header.fragmentCount.toShort())
        buffer.putLong(header.captureTimestampMs)
        buffer.put(header.flags.toByte())
        buffer.putInt(header.sequence)
        return buffer.array()
    }

    /** Splits a full packet into (header AAD, nonce, ciphertext+tag) without touching crypto -
     *  callers decrypt the ciphertext against the returned AAD themselves (keeps this file crypto-
     *  library-agnostic, matching how VideoFrameFraming.kt already separates framing from Crypto). */
    fun parse(packet: ByteArray): ParsedPacket? {
        if (packet.size < HEADER_BYTES + NONCE_BYTES + TAG_BYTES) return null
        val buffer = ByteBuffer.wrap(packet).order(ByteOrder.BIG_ENDIAN)
        val version = buffer.get()
        if (version != VERSION) return null
        val type = buffer.get()
        if (type !in PacketType.all) return null
        val gamingSessionToken = buffer.long
        val frameId = buffer.long
        val fragmentIndex = buffer.short.toInt() and 0xFFFF
        val fragmentCount = buffer.short.toInt() and 0xFFFF
        if (fragmentCount <= 0 || fragmentIndex >= fragmentCount) return null
        val captureTimestampMs = buffer.long
        val flags = buffer.get().toInt() and 0xFF
        val sequence = buffer.int
        val aad = packet.copyOfRange(0, HEADER_BYTES)
        val nonce = packet.copyOfRange(HEADER_BYTES, HEADER_BYTES + NONCE_BYTES)
        val ciphertext = packet.copyOfRange(HEADER_BYTES + NONCE_BYTES, packet.size)
        return ParsedPacket(
            Header(type, gamingSessionToken, frameId, fragmentIndex, fragmentCount, captureTimestampMs, flags, sequence),
            aad,
            nonce,
            ciphertext,
        )
    }

    fun assemble(header: Header, nonce: ByteArray, ciphertext: ByteArray): ByteArray {
        require(nonce.size == NONCE_BYTES)
        val headerBytes = encodeHeader(header)
        return headerBytes + nonce + ciphertext
    }

    /** Splits an encoded video frame's plaintext payload into MTU-safe fragments. Real fragment
     *  encryption/nonce assignment happens per-fragment at the call site (each fragment is its own
     *  independent AEAD-sealed UDP datagram, so losing one fragment never blocks decrypting the
     *  others - there is no cross-fragment authentication dependency). */
    fun splitIntoFragments(payload: ByteArray): List<ByteArray> {
        if (payload.isEmpty()) return listOf(ByteArray(0))
        return payload.toList().chunked(MAX_FRAGMENT_PAYLOAD_BYTES).map { it.toByteArray() }
    }
}
